Move string functions into TextValue

pontusmelke opened the pull request:
Instead of having `StringFunction.scala` implementing string functions we move the implementation of the rest of the string functions down to the `TextValue` implementation. All these methods currently require some form of regular expression support which is why these methods are added on the abstract class using the `value()` method which forces a string to be materialized.

pontusmelke committed:
Moved toLower to TextValue

pontusmelke committed:
Moved toUpper to TextValue

pontusmelke committed:
Moved split to TextValue

pontusmelke committed:
Moved replace to TextValue

pontusmelke committed:
Use StringValue.EMPTY when possible

pontusmelke committed:
Fix failing test

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -44,7 +44,7 @@
     public static final Value ZERO_FLOAT = Values.doubleValue( 0.0 );
     public static final Value ZERO_INT = Values.longValue( 0 );
-    public static final Value MIN_STRING = Values.stringValue( "" );
+    public static final Value MIN_STRING =  StringValue.EMTPY;
```
Extra whitespace

fickludd commented on community/cypher/interpreted-runtime/src/test/scala/org/neo4j/cypher/internal/runtime/interpreted/commands/expressions/StringFunctionsTest.scala:
```diff
@@ -159,4 +158,16 @@ class StringFunctionsTest extends CypherFunSuite {
+    split(null, "hello") should equal(expectedNull)
+    split(null, null) should equal(expectedNull)
+    intercept[CypherTypeException](split(1024, 10))
+  }
```
As these TextValue methods are now moved to the values module, we also have to move/add tests for this functionality to the values module.  If we don't we could leave these methods untested when we remove the interpreted runtime for example.

pontusmelke replied:
Added more tests to `TextValueSpecification` plus added a new test `TextValueTest` that more or less test the same thing as this test.

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/StringValue.java:
```diff
@@ -56,6 +56,39 @@ public boolean equals( String x )
+        String asString = value();
+        //Cypher has different semantics for the case where the separator
+        //is exactly the value, in cypher we expect two empty arrays
+        //where as java returns an empty array
```
Thank you for adding this comment, that is a quite surprising behavior otherwise.

fickludd reviewed (changes requested):
This change make a lot of sense. I have two things that I would like to have fixed before merging:
1) remove that binary jfr recording please :)
2) make test follow the implementation into the values module

pontusmelke committed:
More tests for TextValue

Added more tests and some minor cleanup.
Also discovered a couple of bugs:
- comparing two utf8StringValue with offsets
- compareTo on empty string was not returning length difference as other
  implementations

pontusmelke commented:
@fickludd 
- 1 whoops!
- 2 added more tests

pontusmelke committed:
Dedicated utf8 implementation of reverse

pontusmelke committed:
Move default implementations

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java:
```diff
@@ -0,0 +1,141 @@
+    {
+        assertThat( value.apply( "hello" ).replace( "l", "w" ), equalTo( value.apply( "hewwo" ) ) );
+        assertThat( value.apply( "hello" ).replace( "ell", "ipp" ), equalTo( value.apply( "hippo" ) ) );
+        assertThat( value.apply( "hello" ).replace( "a", "x" ), equalTo( value.apply( "hello" ) ) );
```
Worth putting in replacing with empty string?

pontusmelke replied:
done below

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java:
```diff
@@ -0,0 +1,141 @@
+        assertThat( value.apply( "0123456789" ).substring( 5 ), equalTo( value.apply( "56789" ) ) );
+        assertThat( value.apply( "0123456789" ).substring( 15 ), equalTo( StringValue.EMTPY ) );
+        assertThat( value.apply( "\uD83D\uDE21\uD83D\uDE21\uD83D\uDE21" ).substring( 1, 1 ),
+                equalTo( value.apply( "\uD83D\uDE21" ) ) );
```
This assert is not very powerful. Because the same char is repeated one cannot tell which one was selected.

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java:
```diff
@@ -0,0 +1,141 @@
+    {
+        assertThat( value.apply( "HELLO" ).toLower(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "Hello" ).toLower(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "hello" ).toLower(), equalTo( value.apply( "hello" ) ) );
```
Could try empty string?

pontusmelke replied:
done below

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java:
```diff
@@ -0,0 +1,141 @@
+    {
+        assertThat( value.apply( "HELLO" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( "Hello" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( "hello" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
```
empty string?

pontusmelke replied:
below

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java:
```diff
@@ -0,0 +1,141 @@
+        assertThat( value.apply( "" ).reverse(), equalTo( StringValue.EMTPY ) );
+        assertThat( value.apply( " L" ).reverse(), equalTo( value.apply( "L " ) ) );
+        assertThat( value.apply( "\r\n" ).reverse(), equalTo( value.apply( "\n\r" ) ) );
+        assertThat( value.apply( "\uD801\uDC37" ).reverse(), equalTo( value.apply( "\uD801\uDC37" ) ) );
```
We cannot see it actually reversing for UTF here. It is valuable to test the single char, but maybe also try a real word?

pontusmelke replied:
What do you mean by _reversing for UTF_? The last test is testing that we keep the code point `\uD801\uDC37` intact.

fickludd replied:
Aye, but if you have two code points, they should be in reverse order?

pontusmelke replied:
No it is 𐐷,  http://unicode.scarfboy.com/?s=U%2B10437. \uD801 is a surrogate

fickludd replied:
Ok, I'm not communicating well.

I understand that the current test check that the single code point `\uD801\uDC37` is not reversed to `\uDC37\uD801`, which would be a completely different unicode thing.

What I'm trying to suggest is to also add a multi-codepoint unicode string, to make sure that we actually can reverse these. E.g. `\uDC37\uD801\uDC2D\uD801 => \uDC2D\uD801\uDC37\uD801` or something.

pontusmelke replied:
Something like this?

fickludd replied:
👍

fickludd commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/TextValueSpecification.scala:
```diff
@@ -68,6 +68,40 @@ object TextValueSpecification extends Properties("TextValue") with Configuration
+      value.split(find) == utf8.split(find) && value.split(find) == stringArray("", "")
+    }
+
+  }}
```
I **do** appreciate this randomized testing!

fickludd reviewed (changes requested):
It's great that tests were added to the values module. I just have some nitpicks and suggestions on some of the assertions (which are quite good in general), and then we are done =).

pontusmelke commented:
@fickludd I will add more tests but I was under the impression that you wanted me to port the existing tests? Just to make sure you have seen it, we also have the scalacheck test in `TestValueSpecification`, that test actually gives me more confidence than just adding a few more cases in this file.

fickludd commented:
@pontusmelke I saw `TestValueSpecification` and I agree that those generated test give a lot of credibility to the implementation. It's just annoying that they're at the other end of the code base (in cypher acceptance tests). I would prefer if the tests in the module of the implementation guaranteed it's correctness (at least for something basic like values). 

But if you do not agree we can merge it as is. This is a big improvement compared to before.
